src/server/gc_worker.rs

```diff
@@ -1223,20 +1204,38 @@ impl<E: Engine> GCWorker<E> {
         Ok(())
     }
 
+    /// Check whether GCWorker is busy. If busy, callback will be invoked with an error that
+    /// indicates GCWorker is busy; otherwise, return a new callback that invokes the original
+    /// callback as well as decrease the scheduled task counter.
+    fn check_is_busy<T: 'static>(&self, callback: Callback<T>) -> Option<Callback<T>> {
+        if self.scheduled_tasks.fetch_add(1, atomic::Ordering::SeqCst) > GC_MAX_EXECUTING_TASKS {
+            self.scheduled_tasks.fetch_sub(1, atomic::Ordering::SeqCst);
+            callback(Err(Error::from(ErrorInner::GCWorkerTooBusy)));
+            return None;
+        }
+        let scheduled_tasks = Arc::clone(&self.scheduled_tasks);
+        Some(Box::new(move |r| {
+            scheduled_tasks.fetch_sub(1, atomic::Ordering::SeqCst);
+            callback(r);
+        }))
+    }
+
     pub fn async_gc(
         &self,
         ctx: Context,
         safe_point: TimeStamp,
         callback: Callback<()>,
     ) -> Result<()> {
         KV_COMMAND_COUNTER_VEC_STATIC.gc.inc();
-        self.worker_scheduler
-            .schedule(GCTask::GC {
-                ctx,
-                safe_point,
-                callback,
-            })
-            .or_else(handle_gc_task_schedule_error)
+        self.check_is_busy(callback).map_or(Ok(()), |callback| {
+            self.worker_scheduler
+                .schedule(GCTask::GC {
```

Comment:
Does `schedule` execute the callback immediately? If not then I think the callback on lie 1219 needs to come before the `fetch_sub` or we could keep scheduling tasks past the limit.

Reply:
Sorry I didn't understand.. How to past the limit?

Reply:
Because you decrease `schedule_tasks` before invoking the callback.

Reply:
Did  you mean that another task may come before finishing executing the callback? If so I'll change it, although I think that doesn't really matter since the callback is usually very short.

Reply:
I'm fine with it. And nrc is on holiday, you need to ask for another one for review.